### DISCUSSION TITLE
Absolute Zero

### DISCUSSION THREAD

**Discussion by esabol** [Perl-Oshun/oshun#34] [General]
This is a very, very trivial thing, but I was reading the README.md and saw your Celsius and Fahrenheit examples. It was surprising to me that the Celsius range started at -273.14. Absolute zero is -273.15 °C, of course. Did you intentionally make the lower range -273.14 because absolute zero is not physically achievable? But -273.141 °C is achievable (although difficult, I'm sure).

https://en.wikipedia.org/wiki/Absolute_zero

Quasi-related, have you given any thought to supporting a syntax for open-ended interval numerical ranges like (a,b] or even ]a,b[?

https://en.wikipedia.org/wiki/Interval_(mathematics)#Including_or_excluding_endpoints

**Comment by happy-barney** [comment]
`Range` will be interesting add-on, definitely worth of discussions, but beyond MVP, my feeling is even beyond milestone after mvp

**Comment by Ovid** [comment]
There is some work for ranges [discussed in parameterized checks](https://github.com/Perl-Oshun/oshun/wiki/Initial-Specification#parameterized-checks). We went back and forth on this _repeatedly_. It was hard 😃 [The parameterized arrays](https://github.com/Perl-Oshun/oshun/wiki/Initial-Specification#the-parameterized-array-check) and [parameterized nums](https://github.com/Perl-Oshun/oshun/wiki/Initial-Specification#the-parameterized-num-check) give good examples.

**Reply by Ovid** [reply]
Oh, and the celsius thing was just a booboo on my part. Nice catch!

**Comment by Silyua** [comment]
I mean its true

**Comment by duncand** [comment]
Ranges should be actual value types which can be used in checks but also as actual data, rather than special syntax for checks.  And then they can be used in checks in the same way one can have a check like "value must be a member of this array".  As such, the whole range thing is really orthogonal to Oshun.

**Reply by zmughal** [reply]
After writing this <https://github.com/orgs/Perl-Apollo/discussions/19#discussioncomment-6770070>, I can see this work because if one were to break down a numerical type like `INT` the way I am proposing `:isa(IV | UV | INT_LIKE)`, then it would be very annoying to have to add range parameters and to each of those and pass them down to each part of the union:

```
:isa(IV[$MIN, $MAX] | UV[$MIN, $MAX] | INT_LIKE[$MIN, $MAX])
```

when instead one could do:

```
:isa(INT & Interval[$MIN, $MAX])
```

and also long as the value we are checking implements numerical comparison operators, then it should work. This could be refined easily:

```
:isa(IV     & Interval[$MIN, $MAX])
:isa(U8     & Interval[$MIN, $MAX])
:isa(BigInt & Interval[$MIN, $MAX])
```

depending on the application and could even use the fact that something like `IV` itself has machine-type specific limits (`IV_MAX`) which could be metadata applied to the type. Then part of how the algebra of how `Interval` combines with a numerical check is to see if `$MIN` and `$MAX` are valid. (edit: like <https://en.cppreference.com/w/cpp/types/numeric_limits> in C++, <https://docs.rs/num/latest/num/trait.Bounded.html> in Rust, Bounded in Haskell <https://www.haskell.org/onlinereport/haskell2010/haskellch6.html> to give a range of low and high level languages).

**Reply by druud** [reply]
Which makes me want
`
:isa( Int . gte($MIN) . lt($MAX) )
`
(with gte and lt numerical, because of the "base type")

**Reply by happy-barney** [reply]
@druud what about this ?
```
Int[ max_inclusive ($MIN), min_exclusive ($MAX) ]
```

**Reply by zmughal** [reply]
Or, if we had a way to treat operators as functions (which could be curried), we could do:

```
INT & BinaryOp[ '>=' , $MIN ] & BinaryOp[ '<' , $MAX ]
```

This could be done via a look-up table and something like @tobyink's [`Hydrogen`](https://metacpan.org/pod/Hydrogen::Topic::Number#ge(-$num-)). The reason I'm emphasizing breaking  up the types into simpler rules that can be built up together is that a tool can reason about them without having to know how to parse Perl expressions with proper precedence (which is the difficult part of the grammar that Damian's `PPR` and Kennedy's `PPI` both side step). I am actually working on a mix of Perl and Prolog right now to demonstrate the initial feasibility of this kind of specification of checks.

**Reply by happy-barney** [reply]
@zmughal intention vs implementation ... I prefer to name what I want to achieve.
Also with my approach context of `[...]` is Oshun grammar whereas context of `(...)` is Perl grammar.

**Reply by zmughal** [reply]
Yeah, both are valid and if I were writing a tool for doing extended checking, I don't think it would be too much work to support both to map to the same bit of logic. I would probably write a library of user-defined checks that implement `BinaryOp` and maybe more for other mathematical operations (especially for PDL which will be interesting because it includes shape as a parameter --- which is shares with APL).

**Reply by druud** [reply]
For building up config (and to me a check is similar), I like a chained syntax, where each link refines, and then returns, its input. The '&' approach is similar, if it has some localised $CHECK object that it can mutate at will.
